File: core/src/settings_channel.cpp
/*
 * Copyright © Windower Dev Team
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation files
 * (the "Software"),to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

#include "settings_channel.hpp"

#include "handle.hpp"
#include "utility.hpp"

#include <windows.h>

#include <pugixml.hpp>

#include <array>
#include <cstddef>
#include <cstdint>
#include <iomanip>
#include <sstream>
#include <string>
#include <system_error>
#include <vector>

windower::settings_channel::settings_channel() noexcept
{
    std::wostringstream s;
    s << LR"(Windower.Settings)";
    s << L'[' << std::hex << std::uppercase << std::setw(8)
      << std::setfill(L'0') << ::GetCurrentProcessId() << L']';
    s << LR"(.Data)";
    auto const name = s.str();

    handle const file =
        ::OpenFileMappingW(FILE_MAP_ALL_ACCESS, FALSE, name.c_str());
    if (!file)
    {
        throw_system_error();
    }

    auto const* ptr = ::MapViewOfFile(file, FILE_MAP_READ, 0, 0, 0);
    auto const view = static_cast<std::uint64_t const*>(ptr);
    if (!view)
    {
        throw_system_error();
    }

    auto const size = *view;
    m_document.load_buffer(std::next(view), gsl::narrow<std::size_t>(size));
    m_root = m_document.child("settings");
}

windower::settings_channel::~settings_channel()
{
    std::wostringstream s;
    s << LR"(Windower.Settings)";
    s << L'[' << std::hex << std::uppercase << std::setw(8)
      << std::setfill(L'0') << ::GetCurrentProcessId() << L']';
    s << LR"(.Flag)";
    auto const name = s.str();

    handle const flag = ::OpenEventW(EVENT_MODIFY_STATE, false, name.c_str());
    if (!flag)
    {
        fail_fast();
    }

    ::SetEvent(flag);
}

auto windower::settings_channel::get(
    u8zstring_view key, std::u8string_view def) const -> std::u8string
{
    auto key_ptr = to_zstring_view(key).c_str();
    if (auto const node = m_root.child(key_ptr).text())
    {
        std::string_view const view = node.get();
        return {view.begin(), view.end()};
    }
    return std::u8string{def};
}

auto windower::settings_channel::get(
    u8zstring_view key, char8_t const* def) const -> std::u8string
{
    return get(key, std::u8string_view{def});
}

auto windower::settings_channel::get(u8zstring_view key, bool def) const -> bool
{
    auto key_ptr = to_zstring_view(key).c_str();
    return m_root.child(key_ptr).text().as_bool(def);
}

auto windower::settings_channel::get(u8zstring_view key, float def) const
    -> float
{
    auto key_ptr = to_zstring_view(key).c_str();
    return m_root.child(key_ptr).text().as_float(def);
}

auto windower::settings_channel::get(u8zstring_view key, double def) const
    -> double
{
    auto key_ptr = to_zstring_view(key).c_str();
    return m_root.child(key_ptr).text().as_double(def);
}

auto windower::settings_channel::get(u8zstring_view key, char def) const -> char
{
    auto key_ptr = to_zstring_view(key).c_str();
    if constexpr (std::is_signed_v<char>)
    {
        return gsl::narrow_cast<char>(m_root.child(key_ptr).text().as_int(def));
    }
    else
    {
        return gsl::narrow_cast<char>(
            m_root.child(key_ptr).text().as_uint(def));
    }
}

auto windower::settings_channel::get(u8zstring_view key, signed char def) const
    -> signed char
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<signed char>(
        m_root.child(key_ptr).text().as_int(def));
}

auto windower::settings_channel::get(u8zstring_view key, short def) const
    -> short
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<short>(m_root.child(key_ptr).text().as_int(def));
}

auto windower::settings_channel::get(u8zstring_view key, int def) const -> int
{
    auto key_ptr = to_zstring_view(key).c_str();
    return m_root.child(key_ptr).text().as_int(def);
}

auto windower::settings_channel::get(u8zstring_view key, long def) const
    -> long int
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<long>(m_root.child(key_ptr).text().as_llong(def));
}

auto windower::settings_channel::get(u8zstring_view key, long long def) const
    -> long long int
{
    auto key_ptr = to_zstring_view(key).c_str();
    return m_root.child(key_ptr).text().as_llong(def);
}

auto windower::settings_channel::get(
    u8zstring_view key, unsigned char def) const -> unsigned char
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<unsigned char>(
        m_root.child(key_ptr).text().as_uint(def));
}

auto windower::settings_channel::get(
    u8zstring_view key, unsigned short def) const -> unsigned short
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<unsigned short>(
        m_root.child(key_ptr).text().as_uint(def));
}

auto windower::settings_channel::get(u8zstring_view key, unsigned def) const
    -> unsigned
{
    auto key_ptr = to_zstring_view(key).c_str();
    return m_root.child(key_ptr).text().as_uint(def);
}

auto windower::settings_channel::get(
    u8zstring_view key, unsigned long def) const -> unsigned long
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<unsigned long>(
        m_root.child(key_ptr).text().as_llong(def));
}

auto windower::settings_channel::get(
    u8zstring_view key, unsigned long long def) const -> unsigned long long
{
    auto key_ptr = to_zstring_view(key).c_str();
    return gsl::narrow_cast<unsigned long long>(
        m_root.child(key_ptr).text().as_llong(def));
}

File: launcher/src/Core/SettingsChannel.cs
/*
 * Copyright © Windower Dev Team
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation files
 * (the "Software"),to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge,
 * publish, distribute, sublicense, and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

namespace Windower.Core
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Diagnostics.CodeAnalysis;
    using System.Globalization;
    using System.IO;
    using System.IO.MemoryMappedFiles;
    using System.Linq;
    using System.Threading;
    using System.Threading.Tasks;
    using System.Xml.Linq;

    using static System.FormattableString;

    /// <summary>
    /// Represents a settings communications channel with an external process.
    /// </summary>
    public class SettingsChannel : IDisposable
    {
        /// <summary>
        /// The disposed flag
        /// </summary>
        private bool disposed = false;

        /// <summary>
        /// The shared memory file handle
        /// </summary>
        private MemoryMappedFile file;

        /// <summary>
        /// The completion flag event
        /// </summary>
        private EventWaitHandle flag;

        /// <summary>
        /// Initializes a new instance of the <see cref="SettingsChannel"/> class.
        /// </summary>
        [SuppressMessage("Microsoft.Design", "CA1006")]
        public SettingsChannel(Process process, IEnumerable<KeyValuePair<string, object>> settings)
        {
            if (process == null)
            {
                throw new ArgumentNullException(nameof(process));
            }

            if (settings == null)
            {
                throw new ArgumentNullException(nameof(settings));
            }

            var dataName = Invariant($"Windower.Settings[{process.Id:X8}].Data");
            var flagName = Invariant($"Windower.Settings[{process.Id:X8}].Flag");

            var document = new XDocument(
                new XElement("settings",
                    from pair in settings
                    select new XElement(pair.Key, Unwrap(pair.Value))));

            using (var buffer = new MemoryStream())
            {
                document.Save(buffer, SaveOptions.DisableFormatting);

                file = MemoryMappedFile.CreateNew(dataName, buffer.Length + sizeof(long));
                using (var stream = file.CreateViewStream(0, 0))
                {
                    stream.Write(BitConverter.GetBytes(buffer.Length), 0, sizeof(long));
                    buffer.WriteTo(stream);
                }
            }

            flag = new EventWaitHandle(false, EventResetMode.ManualReset, flagName);
        }

        /// <summary>
        /// Creates an asynchronous task that completes when the external process has signaled that it is finished loading
        /// settings.
        /// </summary>
        public Task FinishAsync(CancellationToken token) =>
            Task.Run(() =>
            {
                do
                {
                    token.ThrowIfCancellationRequested();
                }
                while (!flag.WaitOne(100));
            }, token);

        /// <summary>
        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Releases unmanaged and - optionally - managed resources.
        /// </summary>
        /// <param name="disposing">
        /// <value>true</value> to release both managed and unmanaged resources; <value>false</value> to release only
        /// unmanaged resources.
        /// </param>
        protected virtual void Dispose(bool disposing)
        {
            if (!disposed && disposing)
            {
                file.Dispose();
                flag.Dispose();
            }

            disposed = true;
        }

        private static object Unwrap(object value)
        {
            if (value is Enum e)
            {
                return Convert.ChangeType(e, Enum.GetUnderlyingType(e.GetType()), CultureInfo.InvariantCulture);
            }

            return value;
        }
    }
}
